pkg/client/testing/core/fixture.go

```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+		// not on interfaces, as some interfaces are identical
+		// (e.g. UpdateAction and CreateAction), so if we use them,
+		// updates and creates end up matching the same case branch.
+		switch typedAction := action.(type) {
+
+		case ListActionImpl:
 			kind.Kind += "List"
-			resource, err := o.Kind(kind, "")
-			return true, resource, err
+			obj, err := registry.Get(kind, ns, "")
+			return true, obj, err
 
-		case GetAction:
-			resource, err := o.Kind(kind, castAction.GetName())
-			return true, resource, err
+		case GetActionImpl:
+			obj, err := registry.Get(kind, ns, typedAction.GetName())
+			return true, obj, err
 
-		case DeleteAction:
-			resource, err := o.Kind(kind, castAction.GetName())
-			return true, resource, err
+		case CreateActionImpl:
+			objMeta, err := meta.Accessor(typedAction.GetObject())
+			if err != nil {
+				return true, nil, err
+			}
+			if typedAction.GetSubresource() == "" {
+				err = registry.Add(typedAction.GetObject())
+			} else {
+				// Handle subresource creation as an update on the
+				// enclosing resource.
+				err = registry.Update(typedAction.GetObject())
+			}
+			if err != nil {
+				return true, nil, err
+			}
+			obj, err := registry.Get(kind, ns, objMeta.GetName())
+			return true, obj, err
 
-		case CreateAction:
-			accessor, err := meta.Accessor(castAction.GetObject())
+		case UpdateActionImpl:
+			objMeta, err := meta.Accessor(typedAction.GetObject())
+			if err != nil {
+				return true, nil, err
+			}
+			err = registry.Update(typedAction.GetObject())
 			if err != nil {
 				return true, nil, err
 			}
-			resource, err := o.Kind(kind, accessor.GetName())
-			return true, resource, err
+			obj, err := registry.Get(kind, ns, objMeta.GetName())
+			return true, obj, err
 
-		case UpdateAction:
-			accessor, err := meta.Accessor(castAction.GetObject())
+		case DeleteActionImpl:
+			err := registry.Delete(kind, ns, typedAction.GetName())
 			if err != nil {
 				return true, nil, err
 			}
-			resource, err := o.Kind(kind, accessor.GetName())
-			return true, resource, err
+			return true, nil, nil
 
 		default:
 			return false, nil, fmt.Errorf("no reaction implemented for %s", action)
 		}
 	}
 }
 
-// AddObjectsFromPath loads the JSON or YAML file containing Kubernetes API resources
-// and adds them to the provided ObjectRetriever.
-func AddObjectsFromPath(path string, o ObjectRetriever, decoder runtime.Decoder) error {
-	data, err := ioutil.ReadFile(path)
-	if err != nil {
-		return err
-	}
-	data, err = yaml.ToJSON(data)
-	if err != nil {
-		return err
-	}
-	obj, err := runtime.Decode(decoder, data)
-	if err != nil {
-		return err
-	}
-	if err := o.Add(obj); err != nil {
-		return err
-	}
-	return nil
-}
-
-type objects struct {
-	types   map[string][]runtime.Object
-	last    map[string]int
+type registry struct {
 	scheme  ObjectScheme
 	decoder runtime.Decoder
+	mu      sync.RWMutex
+	objects map[string][]runtime.Object
```

Comment:
I haven't looked carefully, but I think we can get rid of GVK in most places, but we still need GVK where we called [o.scheme.New()](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/testing/core/fixture.go#L161) to get List objects and empty objects.

I'm keen on getting rid of calls to mapper.Kind() because that won't work once we convert controllers to use versioned clientsets, because mapper.Kind() always return the internal kind, while the controllers will expect to receive versioned Lists. Anyway, this is a separate issue, use GVK where you have to and we'll deal with it later.